internal/engine/fswatch/watchmanager.go

```diff
@@ -237,65 +162,143 @@ func (w *WatchManager) OnChange(ctx context.Context, st store.RStore) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 
-	setup, teardown := w.diff(ctx, st)
-
-	// setup the watch first, to avoid a gap in coverage between setup and
-	// teardown. it's ok if we get a file event twice.
-	newWatches := make(map[model.TargetID]targetNotifyCancel)
-	for _, target := range setup {
-		logger := store.NewLogActionLogger(ctx, st.Dispatch)
-		ignore, err := w.createIgnoreMatcher(target)
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
-		}
+	state := st.RLockState()
+	defer st.RUnlockState()
 
-		watcher, err := w.fsWatcherMaker(target.Dependencies(), ignore, logger)
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
-		}
+	if !state.EngineMode.WatchesFiles() {
+		return
+	}
+
+	// TODO(milas): how can global ignores fit into the API model more cleanly?
+	newGlobalIgnores := globalIgnores(state)
+	watchable := FileWatchesForManifests(state.Manifests(), newGlobalIgnores)
+	specsToProcess := make(map[types.NamespacedName]filewatches.FileWatchSpec)
+	for _, w := range watchable {
+		specsToProcess[types.NamespacedName{Namespace: w.GetNamespace(), Name: w.GetName()}] = w.Spec
+	}
 
-		err = watcher.Start()
-		if err != nil {
-			st.Dispatch(store.NewErrorAction(err))
-			continue
+	if len(state.ConfigFiles) > 0 {
+		specsToProcess[types.NamespacedName{Name: ConfigsTargetID.String()}] = filewatches.FileWatchSpec{
+			WatchedPaths: state.ConfigFiles,
 		}
+	}
 
-		ctx, cancel := context.WithCancel(ctx)
-		go w.dispatchFileChangesLoop(ctx, target, watcher, st)
-		newWatches[target.ID()] = targetNotifyCancel{target, watcher, cancel}
+	for name, spec := range specsToProcess {
+		if tw, ok := w.targetWatches[name]; ok {
+			// compare against whatever the current watcher is using as a spec to determine if updates are needed
+			if equality.Semantic.DeepEqual(tw.spec, spec) {
+				continue
+			}
+			// since we're holding a reference to the old one, we just replace it in the map and clean up
+			// after the new watcher is successfully running to avoid missing any events
+			if err := w.addOrReplace(ctx, st, name, spec); err != nil {
+				logger.Get(ctx).Debugf("Error updating watch for %q: %v", name.String(), err)
+				continue
+			}
+			w.cleanupWatch(ctx, tw)
+
+			fw := state.FileWatches[name]
```

Comment:
ah, i hit this problem too...

the approach to this that i've been taking is that, when a subscriber sends a Create/Update, it should keep a local map of the last create/updated object, and then use that map to make create/update idempotent

I think the problem with doing it this way is that you're going to run into weird circular data flow issues where this code is reading and writing the same data....

Reply:
Ha! This is iteration 3, but I should probably just go back to the original. First, I had totally duplicated copies of the full `FileWatch` both places as you're suggesting. A second (short-lived) version was to rely entirely on the store to avoid duplicating here, which obviously did not pan out. I agree this final approach is problematic - I was trying to get closer to the K8s model somewhat (i.e. pretend that it's level-triggered).

I'll flip back to that - I think the next big experiment that needs to be done is shifting responsibility for the creation of the `FileWatch` to the resource to move towards our desired model, at which point, hopefully this becomes much more like a real reconciler, at which point it should just _read_ specs to trigger FS watchers, and let the FS watchers do status updates in the background. At that point, shouldn't be any circular weirdness 😵

Reply:
ya, i had a similar experience - it's simpler if the controller that does create/delete is separate from the controller that does UpdateStatus. i think that having the ChangeSummary stuff to tell the controller what changed will also help with this